x-pack/plugins/dataset_quality/server/routes/data_streams/get_malformed_docs.ts

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { ElasticsearchClient } from '@kbn/core/server';
+import { rangeQuery, termQuery } from '@kbn/observability-plugin/server';
+import {
+  DATA_STREAM_DATASET,
+  DATA_STREAM_NAMESPACE,
+  DATA_STREAM_TYPE,
+  _IGNORED,
+} from '../../../common/es_fields';
+import { DataStreamTypes } from '../../types/data_stream';
+import { MalformedDocs } from '../../types/malformed_docs';
+import { createDatasetQualityESClient, wildcardQuery } from '../../utils';
+
+export async function getMalformedDocsPaginated(options: {
+  esClient: ElasticsearchClient;
+  type?: DataStreamTypes;
+  start: number;
+  end: number;
+  datasetQuery?: string;
+  after?: {
+    dataset: string;
+    namespace: string;
+  };
+  prevResults?: MalformedDocs[];
+}): Promise<MalformedDocs[]> {
+  const { esClient, type = 'logs', datasetQuery, start, end, after, prevResults = [] } = options;
+
+  const datasetQualityESClient = createDatasetQualityESClient(esClient);
+
+  const response = await datasetQualityESClient.search({
+    index: '*',
+    size: 0,
+    query: {
+      bool: {
+        ...(datasetQuery
+          ? {
+              should: [
+                ...wildcardQuery(DATA_STREAM_DATASET, datasetQuery),
+                ...wildcardQuery(DATA_STREAM_NAMESPACE, datasetQuery),
+              ],
+              minimum_should_match: 1,
+            }
+          : {}),
+        filter: [...rangeQuery(start, end), ...termQuery(DATA_STREAM_TYPE, type)],
+      },
+    },
+    aggs: {
+      datasets: {
+        composite: {
+          ...(after ? { after } : {}),
+          size: 10000,
+          sources: [
+            { dataset: { terms: { field: DATA_STREAM_DATASET } } },
+            { namespace: { terms: { field: DATA_STREAM_NAMESPACE } } },
+          ],
+        },
+        aggs: {
+          malformed: {
+            filter: {
+              exists: {
+                field: _IGNORED,
+              },
+            },
+          },
+        },
+      },
+    },
+  });
+
+  const currMalformedDocs =
+    response.aggregations?.datasets.buckets.map((bucket) => ({
+      dataset: `${type}-${bucket.key.dataset}-${bucket.key.namespace}`,
+      percentage: (bucket.malformed.doc_count * 100) / bucket.doc_count,
+    })) ?? [];
+
+  const malformedDocs = [...prevResults, ...currMalformedDocs];
+
+  if (response.aggregations?.datasets.after_key) {
+    return await getMalformedDocsPaginated({
```

Comment:
**nit:** unless there is an explicit error handling, `return await`is redundant and you can directly return the promise of `getMalformedDocsPaginated`.

Reply:
Addressed in [53bd80d](https://github.com/elastic/kibana/pull/172462/commits/53bd80df9d82de42a0e466baf57c32e1a4a4d369).